### PR TITLE
[TEST] Don't reinitialize YamlTestClient and RestClient before each single test

### PR DESCRIPTION
**Pull request by javanna**
In the past we ran yaml tests against an internal cluster, which would get restarted after each test failure, hence the client objects needed to eventually be refreshed before each test. That is why we had the initClient method to re-initialize the YamlTestClient in the execution context. We ended up though re-initializing the client unconditionally, which is not needed.

Also, `ESRestTestCase` recreates the `RestClient` against the external cluster before each test, which is not needed given that nothing changes in the external cluster.

This commit removes the `initClient` method from the yaml tests execution context. The YamlTestClient can be eagerly created before the first yaml test runs and then re-used in subsequent tests. Also api calls to check for nodes versions etc. are moved out of YamlTestClient to `ESClientYamlSuiteTestCase`. Also the RestClient is now initialized in `ESRestTestCase` before the first test runs, and kept around afterwards as a static member. 

Basically each subclass of `EsRestTestCase` will have its own `RestClient` instance, but the client will be shared across the different tests within the same class. The yaml test suite is just a special suite, composed of 600+ tests that are loaded from files, which will share the same client instance.

This change should speed tests up as well, as we don't recreate the `RestClient` before each single test, and we don't call `_cat/nodes` either before each single test.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 1]
```diff
@@ -76,66 +77,72 @@
...
     @After
     public final void after() throws Exception {
         wipeCluster();
         logIfThereAreRunningTasks();
-        closeClients();
+    }
+
+    @AfterClass
+    public static void closeClients() throws IOException {
+        IOUtils.close(client, adminClient);
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: do this in a try / finally block to ensure we let statics go?

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 2]
```diff
@@ -76,66 +77,72 @@
...
+                hosts.add(new HttpHost(host, port, getProtocol()));
+            }
+            clusterHosts = unmodifiableList(hosts);
             client = buildClient(restClientSettings());
             adminClient = buildClient(restAdminSettings());
-        } catch (IOException e) {
-            // Wrap the IOException so children don't have to declare a constructor just to rethrow it.
-            throw new RuntimeException("Error building clients", e);
         }
+        assert client != null && adminClient != null && clusterHosts != null;
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: one assert per member is better then you see what's not null :0

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 3]
```diff
@@ -76,66 +77,72 @@
...
-            String host = stringUrl.substring(0, portSeparator);
-            int port = Integer.valueOf(stringUrl.substring(portSeparator + 1));
-            clusterHosts.add(new HttpHost(host, port, getProtocol()));
-        }
-        this.clusterHosts = unmodifiableList(clusterHosts);
-        try {
+            String[] stringUrls = cluster.split(",");
+            List<HttpHost> hosts = new ArrayList<>(stringUrls.length);
+            for (String stringUrl : stringUrls) {
+                int portSeparator = stringUrl.lastIndexOf(':');
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: we do this kind of parsing in several places maybe a util can help at some point? not sure just an idea

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java [hunk 4]
```diff
@@ -151,12 +214,7 @@ protected void afterIfFailed(List<Throwable> errors) {
         }
 
         //sort the candidates so they will always be in the same order before being shuffled, for repeatability
-        Collections.sort(testCandidates, new Comparator<ClientYamlTestCandidate>() {
-            @Override
-            public int compare(ClientYamlTestCandidate o1, ClientYamlTestCandidate o2) {
-                return o1.getTestPath().compareTo(o2.getTestPath());
-            }
-        });
+        Collections.sort(testCandidates, (o1, o2) -> o1.getTestPath().compareTo(o2.getTestPath()));
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java]: 👍

**Review by s1monw — APPROVED** [review]
left some minors LGTM otherwise given tests pass

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java (hunk 2 above)
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: good point

File: modules/transport-netty4/src/test/java/org/elasticsearch/rest/Netty4HeadBodyIsEmptyIT.java [hunk 5]
```diff
@@ -19,5 +19,53 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonMap;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+public class Netty4HeadBodyIsEmptyIT extends ESRestTestCase {
```
Comment (nik9000) [inline modules/transport-netty4/src/test/java/org/elasticsearch/rest/Netty4HeadBodyIsEmptyIT.java]: ++

**Review by nik9000 — APPROVED** [review]

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java (hunk 3 above)
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: yea makes sense I will have a look at that in a follow-up
